locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/JcloudsLocationSecurityGroupEditor.java

```diff
@@ -0,0 +1,303 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.brooklyn.location.jclouds.networking;
+
+import com.google.common.base.Optional;
+import com.google.common.base.Predicate;
+import com.google.common.base.Predicates;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import org.apache.brooklyn.util.exceptions.Exceptions;
+import org.jclouds.aws.AWSResponseException;
+import org.jclouds.compute.ComputeService;
+import org.jclouds.compute.domain.SecurityGroup;
+import org.jclouds.compute.extensions.SecurityGroupExtension;
+import org.jclouds.domain.Location;
+import org.jclouds.net.domain.IpPermission;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Iterator;
+import java.util.Set;
+import java.util.concurrent.Callable;
+
+/**
+ * Utility for manipulating security groups.
+ *
+ * For convenience, constructors take an argument of a {@link ComputeService} and extract from it the
+ * information needed to edit groups. However, not all compute services support the security group extension,
+ * so if the editor is being created for a compute service where it is not known in advance then users
+ * should query {@link #hasServiceSupport()} before calling the methods to edit security groups.
+ */
+public class JcloudsLocationSecurityGroupEditor {
+
+    private static final Logger LOG = LoggerFactory.getLogger(JcloudsLocationSecurityGroupEditor.class);
+    public static final java.lang.String JCLOUDS_PREFIX = "jclouds#";
+
+    private final Location location;
+    private final ComputeService computeService;
+    private final Optional<SecurityGroupExtension> securityApi;
+    private final String locationId;
+    private final Predicate<Exception> isExceptionRetryable;
+
+    /**
+     * Constructor for editor that will retry operations upon exceptions.
+     * @param location JClouds location where security groups will be managed.
+     * @param computeService The JClouds compute service instance.
+     * @param predicate A predicate indicating whether the customiser can retry a request (e.g. to add a security group
+     * or a rule) if the attempted operation throws a Throwable.
+     */
+    public JcloudsLocationSecurityGroupEditor(Location location, ComputeService computeService, Predicate predicate) {
+        this.location = location;
+        this.computeService = computeService;
+        this.locationId = this.computeService.getContext().unwrap().getId();
+        this.securityApi = this.computeService.getSecurityGroupExtension(); // TODO surely check for isPresent else throw?
```

Comment:
Interesting TODO question. I believe it was previously doing this to help in the writing of portable blueprints. You can add a `JcloudsLocationSecurityGroupCustomizer` to a YAML blueprint, and then use it in clouds that do or don't have security groups. For example, if you deployed to SoftLayer which doesn't need to open the security groups then it would still work (these would be no-ops).

However, it could be argued that a silent (except for log.warn) no-op is not necessarily the right thing, particularly if the cloud requires some different mechanism to be used (e.g. the Firewall-as-a-Service that some OpenStack clouds use).

It would seem sensible to have some level of code that truly does expect the securityGroup extension to exist, and fails otherwise. I'd be happy for this code to assume that (particularly given that some methods do `securityApi.get()`, without checking).